### ISSUE TITLE
--no-resume required if snapshot removed after replication interrupted

### ISSUE THREAD

**Issue by jimsalterjrs** [jimsalterjrs/sanoid#304]
If a replication is interrupted in a snapshot which is destroyed prior to the next attempt at replication, syncoid fails completely on that dataset unless --no-resume is specified, making the resume feature unsafe for unattended use.

Example:

````
Resuming interrupted zfs send/receive from data/images/fileserver to data/prod0/images/fileserver (~ UNKNOWN remaining):
cannot resume send: 'data/images/fileserver@autosnap_2018-09-14_13:00:01_hourly' used in the initial send no longer exists
0.00 B 0:00:00 [0.00 B/s] [<=>                                                                                     ]
cannot receive: failed to read from stream
CRITICAL ERROR: /usr/bin/ssh     -S /tmp/syncoid-root-root@cse-prod0-1544743400 root@cse-prod0 ' /sbin/zfs send -t 1-132fbc1035-110-789c636064000310a501c49c50360710a715e5e7a69766a63040017fa0ac627afc153905209b1d495d7e52566a7209030317541d867c5a5a716a0948262b5d9d0522cf86249f5459925a0ca4134a729660d35f920f7145f5e2b2bf19de961b2390e439c1f27989b9a90c0c29892589fa99b989e9a9c5fa699939a9c5a94565a9450e89a525f9c5798905f146068616ba0696ba8626f186c6560606560686f119f9a54539950c700000fabe2e01 | /bin/gzip -3 | /usr/bin/mbuffer -R 550K -q -s 128k -m 16M 2>/dev/null' | /usr/bin/mbuffer  -q -s 128k -m 16M 2>/dev/null | /bin/zcat  | /usr/bin/pv -s 0 |  /sbin/zfs receive -s -F 'data/prod0/images/fileserver' failed: 256 at /usr/local/bin/syncoid line 469.
````
@phreaker0, since you were the one who coded the resume functionality, do you mind taking a crack at fixing this? If ZFS can't resume due to lack of the missing snapshot, syncoid should fall back to --no-resume behavior from the most recent remaining common snapshot.

This is a pretty easy bug to hit for anybody with Sanoid hourlies and daily offsite replication; get interrupted in the middle of an hourly, wait a day, try again... sanoid's already pruned the hourly, and boom, replication is broken until manual intervention.

**Comment by phreaker0** [comment]
@jimsalterjrs will do. I actually already hit that problem and I also started work on it in a local branch but didn't come around to fix it yet. For me it it didn't work with a normal zfs send as it complained that there is a saved state on the target which I had to destroy first (zfs receive -A), so I guess I have to parse the error output of the zfs command to determine the failure.

**Comment by phreaker0** [comment]
Fixed with #309 and #385

**Comment by jjakob** [comment]
Is this related? Syncoid got stuck in an endless loop:

```
run: syncoid -r --compress=none --no-sync-snap --use-hold --sendoptions=Lcw --preserve-properties --iden
tifier=foo_zfsbackup_3 <source and destination names cut>
NEWEST SNAPSHOT: autosnap_2025-03-28_03:45:02_frequently
Sending incremental rpool/data@autosnap_2025-03-25_22:45:02_frequently ... autosnap_2025-03-28_03:45:02_frequently (~ 33 KB):
NEWEST SNAPSHOT: autosnap_2025-03-28_03:45:02_frequently
Sending incremental rpool/data/subvol-101-disk-0@autosnap_2025-03-25_23:00:19_frequently ... autosnap_2025-03-28_03:45:02_frequently (~ 33 KB):
Resuming interrupted zfs send/receive from rpool/data/subvol-102-disk-1 to foo_zfsbackup_3/ml350_rpool/data/subvol-102-disk-1 (~ UNKNOWN remaining):
cannot resume send: 'rpool/data/subvol-102-disk-1@autosnap_2025-03-05_04:00:19_hourly' used in the initial send no longer exists
cannot receive: failed to read from stream
Resuming interrupted zfs send/receive from rpool/data/subvol-102-disk-1 to foo_zfsbackup_3/ml350_rpool/data/subvol-102-disk-1 (~ UNKNOWN remaining):
cannot resume send: 'rpool/data/subvol-102-disk-1@autosnap_2025-03-05_04:00:19_hourly' used in the initial send no longer exists
cannot receive: failed to read from stream
Resuming interrupted zfs send/receive from rpool/data/subvol-102-disk-1 to foo_zfsbackup_3/ml350_rpool/data/subvol-102-disk-1 (~ UNKNOWN remaining):
cannot resume send: 'rpool/data/subvol-102-disk-1@autosnap_2025-03-05_04:00:19_hourly' used in the initial send no longer exists
cannot receive: failed to read from stream

# syncoid --version
/usr/sbin/syncoid version 2.2.0
(Getopt::Long::GetOptions version 2.54; Perl version 5.38.2)
```

Shouldn't 2.2.0 already have the above fixes?
